source/retryable-writes/retryable-writes.rst

```diff
@@ -511,6 +513,15 @@ experience across all drivers.
 Why expose users to server-side errors for unsupported operations?
 ------------------------------------------------------------------
 
+.. todo:: Determine how this should change. Do we end up using whitelists for
+   supported operations (based on wire protocol verison) and avoid injecting
+   transaction IDs on all write ops, or rely on the server to raise an error if
+   unsupported ops would be retried? In both cases, it's possible for users to
+   be unaware that some operations are not supported until some error in prod.
+   Given the MongoClient-wide scope of retryWrites, we likely don't want to
+   prohibit unsupported operations outright (either in the client API or by
+   having the server raise an exception on the first attempt).
+
```

Comment:
> This is a major behavior change compared to current drivers: it reduces lost updates and increases double-updates.

This highlights an interesting problem with having the server only return an error on retry attempts. The risk of double updates calls to mind your [blog post](https://emptysqua.re/blog/how-to-write-resilient-mongodb-applications/) and the "may overcount" risk of blindly retrying non-idempotent updates after transient network errors. IMO it's a deal-breaker.

That leaves two options:

 * Command whitelist
 * Ability to configure `retryWrites` at narrower levels than the client (e.g. database, collection, operation) to allow users to disable it when they need to perform an unsupported operation and vice versa

The second bullet still seems like it supports the goal of "make it possible to turn on retryability for a subset of operations, with minimum code changes", but I'm inferring "subset" as operations executed within a configured driver object (of any type) instead of operations of a supported type (executed from anywhere).